### ISSUE TITLE
[Question] Loss diverging during SAC training

### ISSUE THREAD

**Issue by tfederico** [DLR-RM/stable-baselines3#1372]
### ❓ Question

Hello,

I am trying to use SAC on a custom environment for controlling a robotic hand (I have already successfully used PPO on it), but sometimes the actor/critic loss randomly diverges to very big values

![image](https://user-images.githubusercontent.com/15323014/224100675-211195f2-57b6-406f-ae3e-66fdcc16534a.png)
![image](https://user-images.githubusercontent.com/15323014/224100741-12713af6-1b31-4fef-adf1-abd73e207dde.png)

As you can see I am running a sweep and the only difference between the runs is the seed used for initialization.
I don't think there is an issue with the environment. Have you ever encountered a similar issue? 

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
please fill the issue template completely (notably hyperparameters used).

Probably duplicate of https://github.com/DLR-RM/rl-baselines3-zoo/issues/322

**Comment by tfederico** [comment]
Apologies, here are the hyperparameters

Policy

- `activation_fn=nn.ReLU`
- `net_arch=[1024, 256]`
- `log_std_init=-3`
- `optimizer_kwargs=dict(weight_decay=1.0e-5)`

Model

- `norm_reward=False`
- `norm_obs=True`
- `learning_rate=0.0001`
- `buffer_size=2000000`
- `learning_starts=5000`
- `batch_size=512`
- `tau=0.1`
- `ent_coef=auto_0.1`
- `target_entropy=auto`
- `gamma=0.99`
- `seed=8`

Trained for 20M steps

**Comment by tfederico** [comment]
I also observed something similar (although on a different task) with the following hyperparams

- `batch_size=512`
- `buffer_size=2000000`
- `ent_coef=auto`
- `learning_rate=0.0003`
- `learning_starts=10000` 
- `log_std_init=-3`
- `target_entropy=auto`
- `tau=1`

**Comment by araffin** [comment]
> optimizer_kwargs=dict(weight_decay=1.0e-5)

if you are using `weight_decay`, you should use `ADAMW` instead of `ADAM`, and the current weight decay is quite small compared to the default (see comments in the linked issue to mitigate the problem).

`tau=0.1` is quite aggressive (compared to the default `0.005`, it is 100x bigger)

**Comment by yannbouteiller** [comment]
I have had a similar issue for ages with another implementation of SAC, I believe it comes from SAC itself. I am trying to track it down, but no success thus far. What I noticed in my case is that the (action-)value estimators diverge toward -infinity at some advanced point in training (although in my situation the environment doesn't even have negative rewards). Do you have this as well maybe?

**Comment by tfederico** [comment]
> > optimizer_kwargs=dict(weight_decay=1.0e-5)
> 
> if you are using `weight_decay`, you should use `ADAMW` instead of `ADAM`, and the current weight decay is quite small compared to the default (see comments in the linked issue to mitigate the problem).
> 
> `tau=0.1` is quite aggressive (compared to the default `0.005`, it is 100x bigger)

Thanks, I will try to change the optimizer and will let you know

**Comment by tfederico** [comment]
Just wanted to drop an update: it seems that using `AdamW` instead of `Adam` and tau as default improves the stability, even with very small weight decay. Some of the runs have random spikes but the order of magnitude is much much smaller

![image](https://user-images.githubusercontent.com/15323014/226675345-62899eee-6cd0-44ec-83dd-677faf25c575.png)

**Comment by araffin** [comment]
> Just wanted to drop an update: it seems that using AdamW instead of Adam and tau as default improves the stability

Thanks for the update, could you give us the value used?

**Comment by tfederico** [comment]
I used the default hyperparameters and weight decay 1.0e-5 for AdamW, but I am planning to experiment with the previous hyperparameters and default tau to see if that's the main reason for the algorithm to diverge.
I am also testing default tau and AdamW with 0.0 as weight decay, which so far seems stable so my guess is that algorithm diverges for high values of tau.
